src/components/WalletManagerContext.tsx

```diff
@@ -26,3 +21,68 @@ export const useWalletManager = () => {
 
   return context
 }
+
+export const useWallet = (
+  chainId?: ChainInfo["chainId"]
+): UseWalletResponse => {
+  const {
+    status: _status,
+    error: _error,
+    connectedWallet: _connectedWallet,
+    chainInfoList,
+  } = useWalletManager()
+
+  const [chainIdStatus, setChainIdStatus] = useState<Status>(
+    Status.Initializing
+  )
+  const [chainIdError, setChainIdError] = useState<unknown>()
+  const [chainIdConnectedWallet, setChainIdConnectedWallet] =
+    useState<ConnectedWallet>()
+  useEffect(() => {
+    if (_status !== Status.Connected || !_connectedWallet || !chainId) {
+      // If the initial wallet client is not yet connected, this chainId
+      // cannot be connected to yet and is thus still initializing.
+      setChainIdStatus(Status.Initializing)
+      setChainIdConnectedWallet(undefined)
+      setChainIdError(undefined)
+      return
+    }
+
+    const connect = async () => {
+      setChainIdStatus(Status.Connecting)
+      setChainIdError(undefined)
+
+      const chainInfo = getChainInfo(chainInfoList, chainId)
+
+      const wallet = Wallets.find(
+        ({ type }) => _connectedWallet.walletType === type
+      )
+      // Smoke test, should never happen.
+      if (!wallet) {
+        throw new Error(`Internal error: could not find wallet.`)
+      }
+
+      setChainIdConnectedWallet(
+        // TODO: Cache
+        await getConnectedWalletInfo(
+          wallet,
+          _connectedWallet.walletClient,
+          chainInfo
+        )
+      )
+      setChainIdStatus(Status.Connected)
+    }
+
+    connect().catch((error) => {
+      console.error(error)
+      setChainIdError(error)
+      setChainIdStatus(Status.Errored)
+    })
+  }, [_status, _connectedWallet, chainId, chainInfoList])
+
+  const status = chainId ? chainIdStatus : _status
+  const error = chainId ? chainIdError : _error
+  const connectedWallet = chainId ? chainIdConnectedWallet : _connectedWallet
+
+  return { status, error, ...connectedWallet }
```

Comment:
Just putting this draft together as an idea/suggestion on the network caching layer. What do you think?

```tsx
export const useWallet = (
  chainId?: ChainInfo["chainId"]
): UseWalletResponse => {
  const { status, error, connectedWallet, chainInfoList } = useWalletManager();

  const initializing =
    status !== Status.Connected || !connectedWallet || !chainId || !chainInfoList;

  const {
    data: chainIdConnectedWallet,
    error: chainIdConnectedWalletError,
    isLoading,
    isError,
    isSuccess
  } = useQuery(
    // will re-run the query everytime any of these keys change
    ["@wallet-connection-query", chainId, chainInfoList],
    async () => {
      const chainInfo = selectChainInfo(chainInfoList, chainId);
      const wallet = Wallets.find(
        ({ type }) => connectedWallet.walletType === type
      );

      // Smoke test, should never happen.
      if (!wallet) {
        throw new Error(`Internal error: could not find wallet.`);
      }

      return await getConnectedWalletInfo(
        wallet,
        connectedWallet.walletClient,
        chainInfo
      );
    },
    {
      // don't run the query unless we're ready
      enabled: !initializing
      // run the query only once 
      // and use its cache if available
      refetchOnWindowFocus: false,
      refetchOnMount: false,
      refetchOnReconnect: false,
      retry: false,
    }
  );

  function getChainIdWalletConnectionStatus() {
    if (isLoading) return Status.Connecting;
    if (isError) return Status.Errored;
    if (isSuccess) return Status.Connected;
  }

  const status = chainId ? getChainIdWalletConnectionStatus() : status;
  const error = chainId ? chainIdConnectedWalletError : error;
  const connectedWallet = chainId ? chainIdConnectedWallet : connectedWallet;

  return { status, error, ...connectedWallet };
};

```

Reply:
great example 👍
let's land this PR first and then follow up discussing this caching. i think this a really good example to start this next discussion 🙂

Reply:
Agree with @ebaker 
Looks sick!